compilers/concrete-optimizer/concrete-optimizer/src/optimization/dag/multi_parameters/analyze.rs

```diff
@@ -1,38 +1,412 @@
+        }
+
+        Iter {
+            iter: self.dag.composition.clone().into_iter(),
+            dag: self,
+        }
+    }
+
+    fn get_operator(&self, index: OperatorIndex) -> VariancedDagOperator<'_> {
+        VariancedDagOperator {
+            dag: self,
+            id: index,
+        }
+    }
+
+    fn get_operator_mut(&mut self, index: OperatorIndex) -> VariancedDagOperatorMut<'_> {
+        VariancedDagOperatorMut {
+            dag: self,
+            id: index,
+        }
+    }
+
+    /// Patches the inputs following the composition rules.
+    fn apply_composition_rules(&mut self) {
+        for compo in self.get_composed_inputs_iter_mut() {
+            compo.apply();
+        }
+    }
+
+    /// Propagates the noise downward in the graph.
+    fn forward_noise(&mut self) -> bool {
+        // We save the old variance to compute the diff at the end.
+        let old_variances = self.variances.clone();
+        let nb_partitions = self.partitions.nb_partitions;
+
+        // We loop through the operators and propagate the noise.
+        for mut operator in self.get_operators_iter_mut() {
+            // Inputs are already computed
+            if operator.operator().is_input() {
+                continue;
+            }
+            let max_var = |acc: SymbolicVariance, input: SymbolicVariance| acc.max(&input);
+            // Operator variance will be used to override the noise
+            let mut operator_variance = OperatorVariance::nan(nb_partitions);
+            // We first compute the noise in the partition of the operator
+            operator_variance[operator.partition().instruction_partition] = match operator
+                .operator()
+                .operator
+            {
+                Operator::Input { .. } => unreachable!(),
+                Operator::Lut { .. } => SymbolicVariance::after_pbs(
+                    nb_partitions,
+                    operator.partition().instruction_partition,
+                ),
+                Operator::LevelledOp { manp, .. } => {
+                    let max_var = operator
+                        .get_inputs_iter()
+                        .map(|a| a.variance()[operator.partition().instruction_partition].clone())
+                        .reduce(max_var)
+                        .unwrap();
+                    max_var.after_levelled_op(*manp)
+                }
+                Operator::Dot {
+                    kind: DotKind::CompatibleTensor { .. },
+                    ..
+                } => todo!("TODO"),
+                Operator::Dot {
+                    kind: DotKind::Unsupported { .. },
+                    ..
+                } => panic!("Unsupported"),
+                Operator::Dot {
+                    inputs,
+                    weights,
+                    kind: DotKind::Simple | DotKind::Tensor | DotKind::Broadcast { .. },
+                } => {
+                    let variance_iterator: Box<dyn Iterator<Item = OperatorVariance>> =
```

Comment:
not sure why Box is needed here. But it could be simpler to have 2 cases and avoid that iterator.

Reply:
Yes, it is better, I splitted the arm in two.